File: hooks/scripts/stop-guard.mjs
#!/usr/bin/env node
// stop-guard.mjs — Stop hook
// Blocks stop when pipeline phases are in-progress

import { readFileSync, readdirSync, existsSync } from 'node:fs'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const RUNS_DIR = join(__dirname, '..', '..', 'pipeline_mcp_data', 'runs')

function findLatestRunState() {
  if (!existsSync(RUNS_DIR)) return null

  const runDirs = readdirSync(RUNS_DIR, { withFileTypes: true })
    .filter(d => d.isDirectory())
    .map(d => d.name)

  let latestRun = null
  let latestTime = 0

  for (const dir of runDirs) {
    const statePath = join(RUNS_DIR, dir, 'run-state.json')
    if (!existsSync(statePath)) continue
    try {
      const state = JSON.parse(readFileSync(statePath, 'utf-8'))
      const updatedAt = new Date(state.updated_at).getTime()
      if (updatedAt > latestTime) {
        latestTime = updatedAt
        latestRun = state
      }
    } catch {
      continue
    }
  }

  return latestRun
}

let input = ''
for await (const chunk of process.stdin) {
  input += chunk
}

try {
  const hookInput = JSON.parse(input || '{}')

  // Prevent infinite loops: if stop_hook_active is set, allow stop
  if (hookInput.stop_hook_active) {
    process.exit(0)
  }

  const runState = findLatestRunState()

  if (!runState || runState.status === 'completed' || runState.status === 'failed') {
    // No active run — allow stop
    process.exit(0)
  }

  // Check for in-progress phases
  const activePhases = Object.entries(runState.phases)
    .filter(([, p]) => p.status === 'in_progress')
    .map(([name]) => name)

  if (activePhases.length > 0) {
    const phaseList = activePhases.join(', ')
    process.stdout.write(JSON.stringify({
      systemMessage: `Warning: Pipeline run "${runState.run_id}" has active phases: ${phaseList}. Consider completing or failing these phases before ending the session. Use pipeline_fail_phase to mark incomplete work, or pipeline_complete_phase if work is done.`,
    }))
  }
} catch {
  // On error, allow stop
  process.exit(0)
}

File: hooks/scripts/stop_guard.py
#!/usr/bin/env python3
"""Warn when pipeline phases are still in-progress at stop — Stop hook."""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from hooks.lib.common import find_latest_run_state, read_stdin  # noqa: E402

_ROOT = Path(__file__).resolve().parents[2]
RUNS_DIR = (
    Path(os.environ.get("PIPELINE_MCP_DATA_DIR") or (_ROOT / "pipeline_mcp_data"))
    / "runs"
)


def main() -> None:
    raw_input = read_stdin()

    try:
        hook_input: Any = json.loads(raw_input or "{}")

        # Prevent infinite loops: if stop_hook_active is set, allow stop.
        if hook_input.get("stop_hook_active"):
            sys.exit(0)

        run_state = find_latest_run_state(RUNS_DIR)

        if not run_state or run_state.get("status") in ("completed", "failed"):
            # No active run — allow stop.
            sys.exit(0)

        # Check for in-progress phases.
        active_phases = [
            name
            for name, p in run_state["phases"].items()
            if p.get("status") == "in_progress"
        ]

        if len(active_phases) > 0:
            phase_list = ", ".join(active_phases)
            sys.stdout.write(
                json.dumps(
                    {
                        "systemMessage": (
                            f'Warning: Pipeline run "{run_state.get("run_id")}" '
                            f"has active phases: {phase_list}. Consider "
                            "completing or failing these phases before ending "
                            "the session. Use pipeline_fail_phase to mark "
                            "incomplete work, or pipeline_complete_phase if "
                            "work is done."
                        )
                    }
                )
            )
    except Exception:
        # On error, allow stop.
        sys.exit(0)


if __name__ == "__main__":
    main()
